test/framework/control_plane.go

```diff
@@ -0,0 +1,225 @@
+	Eventually(func() error {
+		err := mgmtClient.Create(ctx, input.Cluster)
+		if err != nil {
+			fmt.Println(err)
+		}
+		return err
+	}, input.CreateTimeout, 10*time.Second).Should(BeNil())
+
+	// create all the machines at once
+	for _, node := range input.Nodes {
+		By("creating an InfrastructureMachine resource")
+		Expect(mgmtClient.Create(ctx, node.InfraMachine)).To(Succeed())
+
+		By("creating a bootstrap config")
+		Expect(mgmtClient.Create(ctx, node.BootstrapConfig)).To(Succeed())
+
+		By("creating a core Machine resource with a linked InfrastructureMachine and BootstrapConfig")
+		Expect(mgmtClient.Create(ctx, node.Machine)).To(Succeed())
+	}
+
+	// Wait for the cluster infrastructure
+	Eventually(func() string {
+		cluster := &clusterv1.Cluster{}
+		key := client.ObjectKey{
+			Namespace: input.Cluster.GetNamespace(),
+			Name:      input.Cluster.GetName(),
+		}
+		if err := mgmtClient.Get(ctx, key, cluster); err != nil {
+			return err.Error()
+		}
+		return cluster.Status.Phase
+	}, input.CreateTimeout, 10*time.Second).Should(Equal(string(clusterv1.ClusterPhaseProvisioned)))
+
+	// wait for all the machines to be running
+	By("waiting for all machines to be running")
+	for _, node := range input.Nodes {
+		Eventually(func() string {
+			machine := &clusterv1.Machine{}
+			key := client.ObjectKey{
+				Namespace: node.Machine.GetNamespace(),
+				Name:      node.Machine.GetName(),
+			}
+			if err := mgmtClient.Get(ctx, key, machine); err != nil {
+				return err.Error()
+			}
+			return machine.Status.Phase
+		}, input.CreateTimeout, 10*time.Second).Should(Equal(string(clusterv1.MachinePhaseRunning)))
+	}
+
+	By("waiting for the workload nodes to exist")
+	Eventually(func() []v1.Node {
+		nodes := v1.NodeList{}
+		By("ensuring the workload client can be generated")
+		err := wait.PollImmediate(10*time.Second, 5*time.Minute, func() (bool, error) {
+			_, err := input.Management.GetWorkloadClient(ctx, input.Cluster.Namespace, input.Cluster.Name)
+			switch {
+			case apierrors.IsNotFound(err):
+				return false, nil
+			case err != nil:
+				return true, err
+			default:
+				return true, nil
+			}
+		})
+		Expect(err).NotTo(HaveOccurred())
+		By("getting the workload client and listing the nodes")
+		workloadClient, err := input.Management.GetWorkloadClient(ctx, input.Cluster.Namespace, input.Cluster.Name)
+		Expect(err).NotTo(HaveOccurred(), "Stack:\n%+v\n", err)
+		Expect(workloadClient.List(ctx, &nodes)).To(Succeed())
+		return nodes.Items
+	}, input.CreateTimeout, 10*time.Second).Should(HaveLen(len(input.Nodes)))
+}
+
+// CleanUp deletes the cluster and waits for everything to be gone.
+// Assertions:
+//   * Deletes Machines
+//   * Deletes MachineSets
+//   * Deletes MachineDeployments
+//   * Deletes KubeadmConfigs
+//   * Deletes InfraCluster
+//   * Deletes InfraMachines
+//   * Deletes InfraMachineTemplates
+//   * Deletes Secrets
+func (input *ControlplaneClusterInput) CleanUp() {
```

Comment:
If we move the provider specific cleanups to the provider then we should rename this function to CleanUpClusterAPIArtifcats(). That will at least be an attempt to make it clear that it will not clean up provider artifacts and that is the responsibility of the provider.